apps/omg_eth/lib/eth/root_chain.ex

```diff
@@ -19,82 +19,91 @@ defmodule OMG.Eth.RootChain do
   """
 
-  alias OMG.API.Crypto
   alias OMG.Eth
 
-  @type contract_t() :: binary | nil
+  @tx_defaults Eth.Defaults.tx_defaults()
 
-  @spec submit_block(binary, pos_integer, pos_integer, Crypto.address_t() | nil, contract_t()) ::
+  @type optional_addr_t() :: <<_::160>> | nil
+
+  @spec submit_block(binary, pos_integer, pos_integer, optional_addr_t(), optional_addr_t()) ::
           {:error, binary() | atom() | map()}
           | {:ok, binary()}
   def submit_block(hash, nonce, gas_price, from \\ nil, contract \\ nil) do
-    contract = contract || Application.get_env(:omg_eth, :contract_addr)
-    from = from || Application.get_env(:omg_eth, :authority_addr)
+    contract = contract || Eth.Encoding.from_hex(Application.get_env(:omg_eth, :contract_addr))
+    from = from || Eth.Encoding.from_hex(Application.get_env(:omg_eth, :authority_addr))
 
+    # NOTE: we're not using any defaults for opts here!
     Eth.contract_transact(
       from,
       contract,
       "submitBlock(bytes32)",
       [hash],
       nonce: nonce,
       gasPrice: gas_price,
+      value: 0,
       gas: 100_000
     )
   end
 
-  def start_deposit_exit(deposit_positon, value, gas_price, from, contract \\ nil) do
-    contract = contract || Application.get_env(:omg_eth, :contract_addr)
+  def start_deposit_exit(deposit_positon, token, value, from, contract \\ nil, opts \\ nil) do
+    opts = opts || Keyword.put_new(@tx_defaults, :gas, 1_000_000)
+
+    contract = contract || Eth.Encoding.from_hex(Application.get_env(:omg_eth, :contract_addr))
 
     Eth.contract_transact(
       from,
       contract,
-      "startDepositExit(uint256,uint256)",
-      [deposit_positon, value],
-      gasPrice: gas_price,
-      gas: 1_000_000
+      "startDepositExit(uint256,address,uint256)",
+      [deposit_positon, token, value],
+      opts
     )
   end
 
-  def start_exit(utxo_position, txbytes, proof, sigs, gas_price, from, contract \\ nil) do
-    contract = contract || Application.get_env(:omg_eth, :contract_addr)
+  def start_exit(utxo_position, txbytes, proof, sigs, from, contract \\ nil, opts \\ nil) do
+    opts = opts || Keyword.put_new(@tx_defaults, :gas, 1_000_000)
+
+    contract = contract || Eth.Encoding.from_hex(Application.get_env(:omg_eth, :contract_addr))
 
     Eth.contract_transact(
       from,
       contract,
       "startExit(uint256,bytes,bytes,bytes)",
       [utxo_position, txbytes, proof, sigs],
-      gasPrice: gas_price,
-      gas: 1_000_000
+      opts
     )
   end
 
-  def deposit(value, from, contract \\ nil) do
-    contract = contract || Application.get_env(:omg_eth, :contract_addr)
-    Eth.contract_transact(from, contract, "deposit()", [], value: value)
+  def deposit(value, from, contract \\ nil, opts \\ @tx_defaults) do
+    opts = Keyword.put(opts, :value, value)
+    contract = contract || Eth.Encoding.from_hex(Application.get_env(:omg_eth, :contract_addr))
+    Eth.contract_transact(from, contract, "deposit()", [], opts)
```

Comment:
maybe is worth to extract this to separate defp functions like:
```
defp get_contract(contract) when contract == nil , do: Eth.Encoding.from_hex(Application.get_env(:omg_eth, :contract_addr))

defp get_contract(contract), do:  contract
```

Reply:
I imported `Eth.Encoding` and it looks much nicer here now